File: trunk/src/java/cast/core/logging/IceAppender.java
/**
 * 
 */
package cast.core.logging;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;

import org.apache.log4j.AppenderSkeleton;
import org.apache.log4j.helpers.LogLog;
import org.apache.log4j.spi.LoggingEvent;

import Ice.Communicator;
import Ice.Identity;
import Ice.ObjectPrx;
import cast.interfaces.LoggingServerPrx;
import cast.interfaces.LoggingServerPrxHelper;

/**
 * A log4j appender to log to a remote Ice server.
 * 
 * @author nah
 * 
 */
public class IceAppender extends AppenderSkeleton {

	private String m_logHost;
	private int m_logPort;
	private String m_logName;
	private String m_logCategory;
	private LoggingServerPrx m_logServer;

	public IceAppender() throws IOException {
		m_logHost = null;
		m_logPort = LoggingServer.DEFAULT_LOGGER_PORT;
		m_logName = LoggingServer.DEFAULT_LOGGER_NAME;
		m_logCategory = LoggingServer.DEFAULT_LOGGER_CATEGORY;

	}

	@Override
	public void activateOptions() {
		m_logServer = getLoggingServer(m_logHost, m_logPort);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see org.apache.log4j.Appender#close()
	 */
	@Override
	public void close() {
		// SocketAppender
	}

	public void setHost(String _host) {
		m_logHost = _host;
	}

	public void setPort(String _port) {
		m_logPort = Integer.parseInt(_port);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see org.apache.log4j.Appender#requiresLayout()
	 */
	@Override
	public boolean requiresLayout() {
		// TODO Auto-generated method stub
		return false;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * org.apache.log4j.AppenderSkeleton#append(org.apache.log4j.spi.LoggingEvent
	 * )
	 */
	@Override
	protected void append(LoggingEvent _event) {
		// LogLog.debug("message: " + _event.getMessage());

		try {
			
			// TODO check efficiency concerns about doing this
			// every time

			ByteArrayOutputStream os = new ByteArrayOutputStream();
			ObjectOutputStream oos = new ObjectOutputStream(os);
			
			oos.writeObject(_event);
			oos.flush();
			m_logServer.logSerialisedEvent(os.toByteArray());

			
			
		} catch (IOException e) {
			LogLog.error("can't create output event data", e);
		}

	}

	private LoggingServerPrx getLoggingServer(String _host, int _port) {

		Identity id = new Identity(m_logName, m_logCategory);

		ObjectPrx base;
		if (_host != null) {
			base = communicator().stringToProxy(
					communicator().identityToString(id) + ":default -h "
							+ _host + " -p " + _port);
		} else {
			base = communicator().stringToProxy(
					communicator().identityToString(id) + ":default -p "
							+ _port);
		}

		if (base == null) {
			throw new RuntimeException(
					"Cannot create proxy to logging server on " + _host);
		}

		LoggingServerPrx prx = LoggingServerPrxHelper.checkedCast(base);
		assert (prx != null);
		return prx;
	}

	private Communicator m_communicator = null;

	private Communicator communicator() {
		if (m_communicator == null) {
			m_communicator = Ice.Util.initialize();
		}
		return m_communicator;
	}

}

File: src/c++/cast/core/IceAppender.cpp
#include "IceAppender.hpp"

#include <Ice/Communicator.h> 
#include <Ice/Initialize.h> 

#include <log4cxx/helpers/optionconverter.h>
#include <log4cxx/helpers/stringhelper.h>
#include <log4cxx/spi/loggingevent.h>
#include <log4cxx/helpers/objectoutputstream.h>
#include <log4cxx/helpers/bytearrayoutputstream.h>


#include "cast/slice/CDL.hpp"
#include "cast/core/CASTUtils.hpp"

using namespace log4cxx;
using namespace log4cxx::helpers;

using namespace cast;
using namespace cast::core::logging;
using namespace cast::interfaces;

using namespace Ice;

IMPLEMENT_LOG4CXX_OBJECT(IceAppender)

namespace cast {
  namespace core {
    namespace logging {
      
      IceAppender::IceAppender() : 
        m_logHost(),
        m_logPort(cast::cdl::LOGGINGPORT) {
        LogLog::setInternalDebugging(true);
      }
      
      IceAppender::~IceAppender() {
        finalize();
      }
      
      void IceAppender::append(const spi::LoggingEventPtr& event, log4cxx::helpers::Pool& pool) {
        LogLog::debug(LOG4CXX_STR("message: " + event->getMessage()));
//        m_logServer->logMessage(event->getMessage());

        
        // TODO check efficiency concerns about doing this
        // every time

        //Ptr is reference counted, so no manual deletions pleaase...
        ByteArrayOutputStreamPtr os = new ByteArrayOutputStream();
        ObjectOutputStreamPtr oos = new ObjectOutputStream(os,pool);
        
        //this block copied from socketappender.h so I guess it might do something important...
        LogString ndcVal;
        event->getNDC(ndcVal);
        event->getThreadName();
        // Get a copy of this thread's MDC.
        event->getMDCCopy();
        
        try {
          event->write(*oos, pool);
          oos->flush(pool);
          m_logServer->logSerialisedEvent(os->toByteArray());
        } catch(std::exception& e) {
          LogLog::warn(LOG4CXX_STR("Problem dispatching serialised log event: "), e);
        }

        
      }

      void IceAppender::setOption(const LogString& option, const LogString& value) {
        if (StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("HOST"), LOG4CXX_STR("host"))) {
          setHost(value);
            LogLog::debug(LOG4CXX_STR("setHost: " + value));
        }
        else if (StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("PORT"), LOG4CXX_STR("port"))) {
          setPort(OptionConverter::toInt(value, m_logPort));
          LogLog::debug(LOG4CXX_STR("setPort: " + value));

        }
        else {
          AppenderSkeleton::setOption(option, value);
        }
      }

      void IceAppender::activateOptions(log4cxx::helpers::Pool& _pool) {
       	m_logServer = getLoggingServer(m_logHost, m_logPort);        
      }

      LoggingServerPrx 
      IceAppender::getLoggingServer(const std::string & _logHost, const int & _logPort) {
        
        CommunicatorPtr ic = Ice::initialize();      
        assert(ic);
        
        Identity id;
        id.name = "CASTLogServer";
        id.category = "LoggingServer";
        
        std::ostringstream serverAddr;
        serverAddr << ic->identityToString(id)
        << ":default ";
        if(!_logHost.empty()) {
          serverAddr << " -h " << _logHost; 
          }
        serverAddr <<  " -p " << _logPort;
        
        ObjectPrx base = ic->stringToProxy(serverAddr.str());
        assert(base);
        
        LoggingServerPrx logServer = LoggingServerPrx::checkedCast(base);
        assert(logServer);
        return logServer;
      }

    }
  }
}